ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java

```diff
@@ -690,22 +691,20 @@ private SchemaKGroupedStream groupByKey(
       final List<FunctionCall> tableFunctions,
       final QueryContext.Stacker contextStacker
   ) {
-    final List<TableFunctionApplier> tableFunctionAppliers = new ArrayList<>();
+    final List<TableFunctionApplier> tableFunctionAppliers = new ArrayList<>(tableFunctions.size());
+    final CodeGenRunner codeGenRunner =
+        new CodeGenRunner(getSchema(), ksqlConfig, functionRegistry);
     for (FunctionCall functionCall: tableFunctions) {
-      final ColumnReferenceExp exp = (ColumnReferenceExp)functionCall.getArguments().get(0);
-      final ColumnName columnName = exp.getReference().name();
-      final ColumnRef ref = ColumnRef.withoutSource(columnName);
-      final OptionalInt indexInInput = getSchema().valueColumnIndex(ref);
-      if (!indexInInput.isPresent()) {
-        throw new IllegalArgumentException("Can't find input column " + columnName);
-      }
+      final Expression expression = functionCall.getArguments().get(0);
+      final ExpressionMetadata expressionMetadata =
```

Comment:
I feel like this doesn't belong here - instead, we should be resolving all complex expressions before running the FlatMapNode. This has a few benefits, the first is the performance on that I had mentioned earlier (we want to evaluate complex expressions once, not per every output row) and the second is that we're not treating arguments to UDTFs any differently than any other expression.

Reply:
Not sure I follow - this code is before the FlatMapper is run.

Reply:
Talked offline, because this PR is blocking lots of others I'm going to let it through as is and come up with a more detailed ticket describing what I mean by the comment.

Reply:
https://github.com/confluentinc/ksql/issues/3697